datafusion/physical-plan/src/dynamic_filters.rs

```diff
@@ -0,0 +1,320 @@
+pub trait DynamicFilterSource: Send + Sync + std::fmt::Debug + 'static {
+    /// Take a snapshot of the current state of filtering, returning a non-dynamic PhysicalExpr.
+    /// This is used to e.g. serialize dynamic filters across the wire or to pass them into systems
+    /// that won't use the `PhysicalExpr` API (e.g. matching on the concrete types of the expressions like `PruningPredicate` does).
+    /// For example, it is expected that this returns a relatively simple expression such as `col1 > 5` for a TopK operator or
+    /// `col2 IN (1, 2, ... N)` for a HashJoin operator.
+    fn snapshot_current_filters(&self) -> Result<Vec<Arc<dyn PhysicalExpr>>>;
+}
+
+#[derive(Debug)]
+pub struct DynamicFilterPhysicalExpr {
+    /// The children of this expression.
+    /// In particular, it is important that if the dynamic expression will reference any columns
+    /// those columns be marked as children of this expression so that the expression can be properly
+    /// bound to the schema.
+    children: Vec<Arc<dyn PhysicalExpr>>,
+    /// Remapped children, if `PhysicalExpr::with_new_children` was called.
+    /// This is used to ensure that the children of the expression are always the same
+    /// as the children of the dynamic filter source.
+    remapped_children: Option<Vec<Arc<dyn PhysicalExpr>>>,
+    /// The source of dynamic filters.
+    inner: Arc<dyn DynamicFilterSource>,
+    /// For testing purposes track the data type and nullability to make sure they don't change.
+    /// If they do, there's a bug in the implementation.
+    /// But this can have overhead in production, so it's only included in tests.
+    data_type: Arc<RwLock<Option<arrow::datatypes::DataType>>>,
+    nullable: Arc<RwLock<Option<bool>>>,
+}
+
+impl std::fmt::Display for DynamicFilterPhysicalExpr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "DynamicFilterPhysicalExpr")
+    }
+}
+
+// Manually derive PartialEq and Hash to work around https://github.com/rust-lang/rust/issues/78808
+impl PartialEq for DynamicFilterPhysicalExpr {
+    fn eq(&self, other: &Self) -> bool {
+        self.current().eq(&other.current())
+    }
+}
+
+impl Eq for DynamicFilterPhysicalExpr {}
+
+impl Hash for DynamicFilterPhysicalExpr {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.current().hash(state)
+    }
+}
+
+impl DynamicFilterPhysicalExpr {
+    pub fn new(
+        children: Vec<Arc<dyn PhysicalExpr>>,
+        inner: Arc<dyn DynamicFilterSource>,
+    ) -> Self {
+        Self {
+            children,
+            remapped_children: None,
+            inner,
+            data_type: Arc::new(RwLock::new(None)),
+            nullable: Arc::new(RwLock::new(None)),
+        }
+    }
+
+    fn current(&self) -> Arc<dyn PhysicalExpr> {
+        let current = if let Ok(current) = self.inner.snapshot_current_filters() {
```

Comment:
I think it is important to describe here in comments why it is `AND`ing the filter together

Specifically, I think the reason is when there are multiple TopK nodes, each contributes a potentially different range as they have different topk heaps.